share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml

```diff
@@ -951,6 +951,31 @@ e4s-cray-sles-build:
   variables:
     SPACK_CI_STACK_NAME: gcp-base
 
+.gcp-publish:
+  stage: publish
+  only:
```

Comment:
I think most of this stuff can get moved into a common publish rules block. The only thing that is actually different is how the public binary key is copied, everything else can be identical.

If I can get this last PR merged (https://github.com/spack/spack/pull/41297) then we can use an entirely common publish block. I had actually forgotten about that PR until just now. However it isn't required for this to work.

```yaml
.publish:
  stage: publish
  # Don't run publish by default
  before_script:
  - . ./share/spack/setup-env.sh
  - spack --version
  - export COPY_SPECS_DIR=${CI_PROJECT_DIR}/jobs_scratch_dir/specs_to_copy
  - m=$(basename "$PUBLISH_MIRROR"); [[ ! -z $PUBLISH_MIRROR ]] && spack mirror add $m $PUBLISH_MIRROR || echo "Using manifest mirror"
  - spack buildcache sync --manifest-glob "${COPY_SPECS_DIR}/*.json" $(basename "$PUBLISH_MIRROR")
  - curl -fLsS https://spack.github.io/keys/spack-public-binary-key.pub -o /tmp/spack-public-binary-key.pub
  # - spack gpg trust /tmp/spack-public-binary-key.pub
  # - spack buildcache push --only-key spack-public-binary-key
  after_script:
  - . ./share/spack/setup-env.sh
  - spack buildcache update-index --keys $(basename "$PUBLISH_MIRROR"
  variables:
    KUBERNETES_CPU_REQUEST: 4000m
    KUBERNETES_MEMORY_REQUEST: 16G
    PUBLIC_KEY_FILE: 
    PUBLISH_MIRROR: "${PROTECTED_MIRROR_PUSH_DOMAIN}/${CI_COMMIT_REF_NAME}"
  rules:
  # Develop
  - if: $CI_COMMIT_REF_NAME == "develop"
  # Release Branch
  - if: $CI_COMMIT_REF_NAME =~ /^releases\/v.*/
  # Snapshot tags and Release tags
  - if: $CI_COMMIT_TAG =~ /^develop-[\d]{4}-[\d]{2}-[\d]{2}$/ || $CI_COMMIT_TAG =~ /^v.*

.gcp-publish:
  extends: [.publish]
  script:
  - gsutil cp $PUBLIC_KEY_FILE "${PUBLISH_MIRROR}/build_cache/_pgp/spack-public-binary-key.pub"

.aws-publish:
  extends: [.publish]
  script:
  - aws s3 cp $PUBLIC_KEY_FILE "${PUBLISH_MIRROR}/build_cache/_pgp/spack-public-binary-key.pub"

gcp-base-publish:
  extends: [.gcp-publish]
  variables:
    PUBLISH_MIRROR: "gs://dwjacobsen/spack_cache/${CI_COMMIT_REF_NAME}/gcp-base"
  depends:
   - gcp-base-generate
```